plugins/logs/plugin.go

```diff
@@ -401,12 +401,48 @@ func (p *Plugin) Start(ctx context.Context) error {
 // Stop stops the plugin.
 func (p *Plugin) Stop(ctx context.Context) {
 	p.logInfo("Stopping decision logger.")
+
+	gracefulDeadline, _ := ctx.Deadline()
+	gracefulShutdownPeriod := gracefulDeadline.Sub(time.Now())
+
+	if p.config.Service != "" && gracefulShutdownPeriod > 0 {
+		p.flushDecisions(context.WithTimeout(ctx, gracefulShutdownPeriod))
+	}
+
 	done := make(chan struct{})
 	p.stop <- done
 	_ = <-done
 	p.manager.UpdatePluginStatus(Name, &plugins.Status{State: plugins.StateNotReady})
 }
 
+func (p *Plugin) flushDecisions(ctx context.Context, cancel context.CancelFunc) {
+	p.logInfo("Flushing decision logs.")
+	defer cancel()
+
+	go func(ctx context.Context, cancel context.CancelFunc) {
+		for {
+			ok, err := p.oneShot(ctx)
+			if err != nil {
+				p.logError("%v.", err)
+			} else if ok {
+				cancel()
+			}
+			// Wait some before retrying, but skip incrementing interval since we are shutting down
+			time.Sleep(1 * time.Second)
+		}
+	}(ctx, cancel)
```

Comment:
So since we're blocking here, what's the use of the goroutine?

Reply:
I'd think that `p.oneShot(ctx)` will fail if the deadline was exceeded... so we can check that, maybe?

Reply:
> So since we're blocking here, what's the use of the goroutine?

I'm not sure what the alternative would look like - since the for loop is blocking would you not need to run it in it's own go routine for the select to be reachable? I'm learning the go concurrency primitives as I code along so if there's an obvious rookie mistake here I'll happily fix it :)